source/comp/markv_codec.cpp

```diff
@@ -0,0 +1,1478 @@
+    use_delimiter_ = false;
+  }
+
+  void AppendBitSequence(const std::string& str) {
+    if (use_delimiter_)
+      Append("-");
+    Append(str);
+    use_delimiter_ = true;
+  }
+
+  void AppendWhitespaces(size_t num) {
+    Append(std::string(num, ' '));
+    use_delimiter_ = false;
+  }
+
+  void NewLine() {
+    Append("\n");
+    use_delimiter_ = false;
+  }
+
+  std::string GetText() const {
+    return ss_.str();
+  }
+
+ private:
+  void Append(const std::string& str) {
+    ss_ << str;
+    // std::cerr << str;
+  }
+
+  std::stringstream ss_;
+
+  // If true a delimiter will be appended before the next bit sequence.
+  // Used to generate outputs like: 1100-0 1110-1-1100-1-1111-0 110-0.
+  bool use_delimiter_ = false;
+};
+
+// Creates spv_text object containing text from |str|.
+// The returned value is owned by the caller and needs to be destroyed with
+// spvTextDestroy.
+spv_text CreateSpvText(const std::string& str) {
+  spv_text out = new spv_text_t();
+  assert(out);
+  char* cstr = new char[str.length() + 1];
+  assert(cstr);
+  std::strncpy(cstr, str.c_str(), str.length());
+  cstr[str.length()] = '\0';
+  out->str = cstr;
+  out->length = str.length();
+  return out;
+}
+
+// Base class for MARK-V encoder and decoder. Contains common functionality
+// such as:
+// - Validator connection and validation state.
+// - SPIR-V grammar and helper functions.
+class MarkvCodecBase {
+ public:
+  virtual ~MarkvCodecBase() {
+    spvValidatorOptionsDestroy(validator_options_);
+  }
+
+  MarkvCodecBase() = delete;
+
+  void SetModel(const MarkvModel* model) {
+    model_ = model;
+  }
+
+ protected:
+  struct MarkvHeader {
+    MarkvHeader() {
+      magic_number = kMarkvMagicNumber;
+      markv_version = GetMarkvVersion();
+      markv_model = 0;
+      markv_length_in_bits = 0;
+      spirv_version = 0;
+      spirv_generator = 0;
+    }
+
+    uint32_t magic_number;
+    uint32_t markv_version;
+    // Magic number to identify or verify MarkvModel used for encoding.
+    uint32_t markv_model;
+    uint32_t markv_length_in_bits;
+    uint32_t spirv_version;
+    uint32_t spirv_generator;
+  };
+
+  explicit MarkvCodecBase(spv_const_context context,
+                          spv_validator_options validator_options)
+      : validator_options_(validator_options),
+        vstate_(context, validator_options_), grammar_(context),
+        model_(GetDefaultModel()) {}
+
+  // Validates a single instruction and updates validation state of the module.
+  spv_result_t UpdateValidationState(const spv_parsed_instruction_t& inst) {
+    return ValidateInstructionAndUpdateValidationState(&vstate_, &inst);
+  }
+
+  // Returns the current instruction (the one last processed by the validator).
+  const Instruction& GetCurrentInstruction() const {
+    return vstate_.ordered_instructions().back();
+  }
+
+  // Returns true if the opcode has a variable number of operands.
+  bool OpcodeHasVariableNumberOfOperands(SpvOp opcode) const {
```

Comment:
I think this does not handle things like OpDecorate which can have differing numbers of operands based on one of its enum values.
E.g.   OpDecorate %target BuiltIn Position
         OpDecorate %target Volatile

Discussed offline, and perhaps it's safer to use a whitelist of instructions that certainly have a fixed number of logical operands.  And for quality-of-compression purposes, you only need to cover the top 20 or whatever until you get the long tail.